lib/pos_consensus.go

```diff
@@ -602,61 +602,30 @@ func (cc *FastHotStuffConsensus) fetchValidatorListsForSafeBlocks(blocks []*MsgD
 	if err != nil {
 		return nil, errors.Errorf("Error creating UtxoView: %v", err)
 	}
 
-	// Fetch the current epoch entry for the committed tip
-	epochEntryAtCommittedTip, err := utxoView.GetCurrentEpochEntry()
-	if err != nil {
-		return nil, errors.Errorf("Error fetching epoch entry for committed tip: %v", err)
-	}
-
-	// Fetch the next epoch entry
-	nextEpochEntryAfterCommittedTip, err := utxoView.simulateNextEpochEntry(epochEntryAtCommittedTip.EpochNumber, epochEntryAtCommittedTip.FinalBlockHeight)
+	// Fetch the validator set for each block
+	validatorSetByBlockHeight, err := utxoView.GetSnapshotValidatorSetsForBlockHeights(blockHeights)
 	if err != nil {
-		return nil, errors.Errorf("Error fetching next epoch entry after committed tip: %v", err)
+		return nil, errors.Errorf("Error fetching validator set for blocks: %v", err)
 	}
 
-	// The input blocks can only be part of the current or next epoch entries.
-	possibleEpochEntriesForBlocks := []*EpochEntry{epochEntryAtCommittedTip, nextEpochEntryAfterCommittedTip}
-
-	// Fetch the validator set at each block
+	// Extract the validator set at each block
 	blocksWithValidatorLists := make([]consensus.BlockWithValidatorList, len(blocks))
 	for ii, block := range blocks {
-		// Find the epoch entry for the block. It'll either be the current epoch entry or the next one.
-		// We add 1 to the block height because we need the validator set that results AFTER connecting the
-		// block to the blockchain, and triggering an epoch transition (if at an epoch boundary).
-		epochEntryForBlock, err := getEpochEntryForBlockHeight(block.Height+1, possibleEpochEntriesForBlocks)
-		if err != nil {
-			return nil, errors.Errorf("Error fetching epoch number for block: %v", err)
-		}
-
-		// Compute the snapshot epoch number for the block. This is the epoch number that the validator set
-		// for the block was snapshotted in.
-		snapshotEpochNumber, err := utxoView.ComputeSnapshotEpochNumberForEpoch(epochEntryForBlock.EpochNumber)
-		if err != nil {
-			return nil, errors.Errorf("error computing snapshot epoch number for epoch: %v", err)
-		}
-
-		var validatorSetAtBlock []*ValidatorEntry
-		var ok bool
-
-		// If the validator set for the block is already cached by the snapshot epoch number, then use it.
-		// Otherwise, fetch it from the UtxoView.
-		if validatorSetAtBlock, ok = validatorSetEntriesBySnapshotEpochNumber[snapshotEpochNumber]; !ok {
-			// We don't have the validator set for the block cached. Fetch it from the UtxoView.
-			validatorSetAtBlock, err = utxoView.GetAllSnapshotValidatorSetEntriesByStakeAtEpochNumber(snapshotEpochNumber)
-			if err != nil {
-				return nil, errors.Errorf("Error fetching validator set for block: %v", err)
-			}
+		validatorSetAtBlock, ok := validatorSetByBlockHeight[block.Height+1]
+		if !ok {
+			return nil, errors.Errorf("Error fetching validator set for block: %v", block)
 		}
 
 		blocksWithValidatorLists[ii] = consensus.BlockWithValidatorList{
```

Comment:
Would it make sense to convert all the ValidatorEntries to ConsensusInterface before the loop? Minor optimization I guess, but if the number of safe blocks is large (m) and the number of validators is also somewhat large (n), we get a O(m*n) with having it in the loop.

Reply:
Yes! Great idea

Reply:
It'll be O(m*n) even if we perform the conversion before the loop and want to do it in a simple way. I feel fine about leaving the conversion as-is since the number of safe blocks will be small (< 10).